### ISSUE TITLE
[HOLD for payment 2023-08-07] [$1000] Preferences- Notification toggle updating inconsistently on devices

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#19640]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Precondition - Login to the same account in Main device (Desktop) and secondary device (Web)

Offline test:
1. Open Desktop app
2. Open preferences (on Web & Desktop)
3. Click "Force offline" (on Desktop)
4. Click 3 times on Notifications toggle (On Desktop)
5. Click on "Force offline" toggle to go online (On Desktop)

Online test:
1. Open Desktop app
2. Open preferences (on Web & Desktop)
4. Click 2 times quickly on the Notifications toggle (On Desktop)
5. See that it updated correctly on web, but the desktop is now wrong

## Expected Result:
Notifications toggle is synced on main and secondary device.

## Actual Result:
Notifications toggle differs on main and secondary device. When user goes online on main device - notification toggle blinks on the secondary device but then goes back to the previous state which differs from the main device state.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.18-2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes 

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://platform.applause.com/services/links/v1/external/0875e549d85a3e7f06f3a10355369edfe2ef01dfd60007c73c53f56e5c845377

https://github.com/Expensify/App/assets/30054992/e50149da-fd68-4a63-b423-1f9d843821a0

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal
**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01240a84e969686e0b</li>
        <li>Upwork Job ID: 1662121552691892224</li>
        <li>Last Price Increase: 2023-06-09</li>
    </ul>
</details>

**Comment by sakluger** [comment]
I was able to reproduce this bug. One other note is that after following the reproduction steps, I refreshed the secondary device, and it synced the notification setting back off to match the primary device.

**Comment by namhihi237** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The notifications toggle is synced on a primary and secondary device.
### What is the root cause of that problem?
The issue here arises when the device is offline and requests are being queued into the persistedRequests array. When the device reestablishes an internet connection, these requests are retrieved and sent out in sequence. After each request is processed, it is removed from the array.

However, a problem occurs when the Notifications toggle is clicked three times, causing identical requests (number 1 and 3) to be queued into the persistedRequests array. Upon processing, the function PersistedRequests.remove(requestToProcess); is invoked, which results in the removal of both request 1 and 3 due to their similarity. Consequently, we lose one request in the process.

https://github.com/Expensify/App/blob/12d2f3c89de6c0271ae90760de816e198d3e9025/src/libs/actions/PersistedRequests.js#L34-L37
### What changes do you think we should make in order to solve the problem?
Instead of deleting all matching requests, you can modify your removal function to delete only the first matching request. This ensures that if there are duplicate requests in the persistedRequests array, only the one currently being processed gets deleted.
```
function remove(requestToRemove) {
    const index = _.findIndex(persistedRequests, (persistedRequest) => _.isEqual(persistedRequest, requestToRemove));
    if (index !== -1) {
        persistedRequests.splice(index, 1);
    }
    Onyx.set(ONYXKEYS.PERSISTED_REQUESTS, persistedRequests);
}
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by danieldoglas** [comment]
That root cause makes sense to me, but I'm curious if this change won't break anything else - We probably remove all of them for a reason. cc @marcaaron, can you confirm?

If that's the case, we might need to create a new property in the queue object to confirm it is a unique request, and only delete the first if it is.

**Comment by danieldoglas** [comment]
@allroundexperts for the error you found, can you confirm if it's a concurrency error? IMO the second device should have received all 3 pusher updates, but it might be out of order. If that's the case, I'm not sure we can fix it the way our pusher updates work today.

**Comment by danieldoglas** [comment]
Oh no, I thought you had clicked 3 times. Yeah, I think we should fix it here as well.

**Comment by danieldoglas** [comment]
done.

**Comment by namhihi237** [comment]
Hi @allroundexperts , I just checked the additional issue with the online, the issue happens when clicking a second time while the first API is not finished. I think we can use Debounce for this.

**Comment by namhihi237** [comment]
Yes, @allroundexperts , I updated (https://github.com/Expensify/App/issues/19640#issuecomment-1565632028)

**Comment by sakluger** [comment]
Hey @allroundexperts mind taking another look?

**Comment by marcaaron** [comment]
Sorry, I actually don't know what I am supposed to confirm... was there a more specific question that I missed? Can someone summarize it?

**Comment by namhihi237** [comment]
Hi @marcaaron as my understanding for my [proposal](https://github.com/Expensify/App/issues/19640#issuecomment-1565632028) update function remove an item after call API from persistedRequests when the internet is back online:
This is current:
```
persistedRequests = _.reject(persistedRequests, (persistedRequest) => _.isEqual(persistedRequest, requestToRemove)); 
```

My updated:
```
const index = _.findIndex(persistedRequests, (persistedRequest) => _.isEqual(persistedRequest, requestToRemove));
    if (index !== -1) {
        persistedRequests.splice(index, 1);
    }
```
What @allroundexperts  want to confirm here is whether there is a reason to delete all the requests same request that just finished running here in the current code.

**Comment by marcaaron** [comment]
I don't really understand the question. My first instinct is that I don't think we should be modifying these requests at all and unsure why we are removing any requests in the first place. 

Can someone provide more context on why we want to do that? 

I would prefer if we bring this discussion to Slack so it can get wider feedback instead of blocking on any confirmation from me...

**Comment by marcaaron** [comment]
Hard to say without a full understanding but it feels like we are not really addressing any root cause and kind of "hacking at the branches" of some larger problem we've identified with persisted requests.

**Comment by sakluger** [comment]
Readded BZ label since I'm OOO next week.

**Comment by greg-schroeder** [comment]
Reaching out to Callstack/SWM

**Comment by ukane-philemon** [comment]
> https://expensify.slack.com/archives/C03UK30EA1Z/p1686328933116559

IMO, the notification sent to the other device is triggered too early. The animation did not finish before the note was sent out. So for the online sync issue, buttons that have animation should complete before the action/notification for it is initiated.

I dunno if I'm making sense 🌚 .

**Comment by lukemorawski** [comment]
Working on the issue

**Comment by lukemorawski** [comment]
The proposal is CS internal review.

**Comment by greg-schroeder** [comment]
Looks like @lukemorawski is still working on this! Is there anything we can do to assist?

**Comment by lukemorawski** [comment]
@greg-schroeder Hey! Any updates on that issue?

**Comment by danieldoglas** [comment]
@greg-schroeder I'll remove myself from this issue until @allroundexperts reviews this, and then another engineer will be assigned. I'm going OOO for 2 weeks, and don't want to block this issue on me!

**Comment by quinthar** [comment]
> This issue from the second part of the ticket is no longer reproducible, on any platform any type of device. Seems that it might have something to do with how BE processed requests.

Can you elaborate on what you mean by "might have something to do with how BE processed requests" -- what is "BE" in this context?

**Comment by lukemorawski** [comment]
> > This issue from the second part of the ticket is no longer reproducible, on any platform any type of device. Seems that it might have something to do with how BE processed requests.
> 
> Can you elaborate on what you mean by "might have something to do with how BE processed requests" -- what is "BE" in this context?

@quinthar The issue was easily reproducible and suddenly it wasn't. Nothing has changed in the FE code on my side, ergo something must've changed on the BE.

**Comment by greg-schroeder** [comment]
Bump @allroundexperts @puneetlath on the above? 🙏

**Comment by puneetlath** [comment]
@abdulrahuman5196 yes that makes sense to me. Let's wait to see if we come up with a general solution in that discussion.

**Comment by puneetlath** [comment]
Still on hold.

**Comment by greg-schroeder** [comment]
Same as above

**Comment by abdulrahuman5196** [comment]
@puneetlath @allroundexperts we where unable to implement the general solution as part of the other issue, since we discussed and decided the issue itself something we won't address.

**Comment by abdulrahuman5196** [comment]
@allroundexperts I would leave to your discretion. The idea we mentioned was also a valid idea, the reason it was not implemented was that the other issue didn't seem significant to solve. 
But you could also choose to implement it here. Anyways will let you choose.

**Comment by puneetlath** [comment]
I agree, let's not worry about stripping duplicate requests. That seems like a can of worms. Just to make sure that I understand, going with the approach of removing the first matching request would effectively mean we handle the requests in order, correct? So if I toggle-on, toggle-off, toggle-on, we'll do all three actions, in order, ending with the toggle being on? 

If so, that makes sense to me.

**Comment by greg-schroeder** [comment]
I'll create a new job when we're ready, they just time out after 30 days

**Comment by greg-schroeder** [comment]
Last update [4 days ago](https://github.com/Expensify/App/pull/22727#issuecomment-1636166607), perhaps @puneetlath can take a look soon!

**Comment by greg-schroeder** [comment]
This shouldn't have been on hold, but we're reverting the PR given it caused 2 deploy blockers, so this will essentially be reset.

**Comment by puneetlath** [comment]
Reminder when paying: this issue had a regression.

**Comment by greg-schroeder** [comment]
Sounds good. Holding for payment and we can close, then.

**Comment by greg-schroeder** [comment]
Not overdue

**Comment by greg-schroeder** [comment]
Will process today as we've reached new payment date

**Comment by greg-schroeder** [comment]
The original upwork job is closed because it's so old, so I'll have to create a new one

**Comment by greg-schroeder** [comment]
Issue Participants:

Issue reported by: Applause - Internal
Contributor: @namhihi237 
C+: @allroundexperts 

Was this issue merged in time to be eligible for the speed bonus? ❌ 
Were their any regressions? ✅ 

Payment summary:

Contributor: @namhihi237 - $500
C+: @allroundexperts - $500

**Comment by greg-schroeder** [comment]
Payments made in Upwork. Just need @JmillsExpensify to confirm payment for C+ and then we can close as checklist isn't required.

**Comment by greg-schroeder** [comment]
Pending newdot payment

**Comment by greg-schroeder** [comment]
Same as above
